File: android/src/main/java/com/jumio/react/JumioModule.java
/*
 * Copyright (c) 2021. Jumio Corporation All rights reserved.
 */

package com.jumio.react;

import android.app.Activity;
import android.content.Intent;

import com.facebook.react.bridge.ActivityEventListener;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.BaseActivityEventListener;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.jumio.defaultui.JumioActivity;
import com.jumio.sdk.credentials.JumioCredentialInfo;
import com.jumio.sdk.result.JumioCredentialResult;
import com.jumio.sdk.result.JumioFaceResult;
import com.jumio.sdk.result.JumioIDResult;
import com.jumio.sdk.result.JumioResult;

import java.util.List;

public class JumioModule extends JumioBaseModule {

    private final static String TAG = "JumioMobileSDK";
    private final static int REQUEST_CODE = 101;

    private final String ERROR_KEY = "EventError";

    JumioModule(ReactApplicationContext context) {
        super(context);
        // Add the listener for `onActivityResult`
    }

    @Override
    public String getErrorKey() {
        return ERROR_KEY;
    }

    private final ActivityEventListener mActivityEventListener = new BaseActivityEventListener() {
        @Override
        public void onActivityResult(Activity activity, int requestCode, int resultCode, Intent data) {
            if (requestCode == REQUEST_CODE) {
                if (data == null) {
                    return;
                }

                final JumioResult jumioResult = (JumioResult) data.getSerializableExtra(JumioActivity.EXTRA_RESULT);

                if (jumioResult != null && jumioResult.isSuccess()) {
                    sendScanResult(jumioResult);
                } else {
                    sendCancelResult(jumioResult);
                }
                reactContext.removeActivityEventListener(mActivityEventListener);
            }
        }
    };

    @ReactMethod
    public void initialize(String authorizationToken, String dataCenter) {
        try {
            if (authorizationToken.isEmpty() || dataCenter.isEmpty()) {
                showErrorMessage("Missing required parameters one-time session authorization token, or dataCenter.");
                return;
            }

            final Intent intent = new Intent(getCurrentActivity(), JumioActivity.class);
            intent.putExtra(JumioActivity.EXTRA_TOKEN, authorizationToken);
            intent.putExtra(JumioActivity.EXTRA_DATACENTER, dataCenter);

//            The following intent extra can be used to customize the Theme of Default UI
//            intent.putExtra(JumioActivity.EXTRA_CUSTOM_THEME, R.style.AppThemeCustomJumio);

            getCurrentActivity().startActivityForResult(intent, REQUEST_CODE);

        } catch (Exception e) {
            showErrorMessage("Error initializing the Jumio SDK: " + e.getLocalizedMessage());
        }
    }

    @ReactMethod
    public void start() {
        try {
            boolean sdkStarted = checkPermissionsAndStart();
            if (sdkStarted) {
                reactContext.addActivityEventListener(mActivityEventListener);
            }
        } catch (Exception e) {
            showErrorMessage("Error starting the Jumio SDK: " + e.getLocalizedMessage());
        }
    }

    private void sendScanResult(final JumioResult jumioResult) {
        final String accountId = jumioResult.getAccountId();
        final List<JumioCredentialInfo> credentialInfoList = jumioResult.getCredentialInfos();

        final WritableMap result = Arguments.createMap();
        final WritableArray credentialsArray = Arguments.createArray();

        if (credentialInfoList != null) {
            if (accountId != null) {
                result.putString("accountId", accountId);
            }
            for (JumioCredentialInfo credentialInfo : credentialInfoList) {
                final JumioCredentialResult jumioCredentialResult = jumioResult.getResult(credentialInfo);
                final WritableMap credentialMap = Arguments.createMap();

                credentialMap.putString("credentialId", credentialInfo.getId());
                credentialMap.putString("credentialCategory", credentialInfo.getCategory().toString());

                if (jumioCredentialResult instanceof JumioIDResult) {
                    final JumioIDResult idResult = (JumioIDResult) jumioCredentialResult;
                    credentialMap.putString("selectedCountry", idResult.getCountry());
                    credentialMap.putString("selectedDocumentType", idResult.getIdType());
                    credentialMap.putString("idNumber", idResult.getDocumentNumber());
                    credentialMap.putString("personalNumber", idResult.getPersonalNumber());
                    credentialMap.putString("issuingDate", idResult.getIssuingDate());
                    credentialMap.putString("expiryDate", idResult.getExpiryDate());
                    credentialMap.putString("issuingCountry", idResult.getIssuingCountry());
                    credentialMap.putString("lastName", idResult.getLastName());
                    credentialMap.putString("firstName", idResult.getFirstName());
                    credentialMap.putString("gender", idResult.getGender());
                    credentialMap.putString("nationality", idResult.getNationality());
                    credentialMap.putString("dateOfBirth", idResult.getDateOfBirth());

                    credentialMap.putString("addressLine", idResult.getAddress());
                    credentialMap.putString("city", idResult.getCity());
                    credentialMap.putString("subdivision", idResult.getSubdivision());
                    credentialMap.putString("postCode", idResult.getPostalCode());
                    credentialMap.putString("placeOfBirth", idResult.getPlaceOfBirth());

                    credentialMap.putString("mrzLine1", idResult.getMrzLine1());
                    credentialMap.putString("mrzLine2", idResult.getMrzLine2());
                    credentialMap.putString("mrzLine3", idResult.getMrzLine3());
                } else if (jumioCredentialResult instanceof JumioFaceResult) {
                    //lowercase
                    credentialMap.putString("passed", String.valueOf(((JumioFaceResult) jumioCredentialResult).getPassed()));
                }
                credentialsArray.pushMap(credentialMap);
            }
            result.putArray("credentials", credentialsArray);
        }
        sendEvent("EventResult", result);
    }

    private void sendCancelResult(final JumioResult jumioResult) {
        if (jumioResult != null && jumioResult.getError() != null) {
            String errorMessage = jumioResult.getError().getMessage();
            String errorCode = jumioResult.getError().getCode();
            sendErrorObject(errorCode, errorMessage);
        } else {
            showErrorMessage("There was a problem extracting the scan result");
        }
    }
}

File: android/src/main/java/com/jumio/react/JumioModule.kt
package com.jumio.react

import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.BaseActivityEventListener
import com.facebook.react.bridge.Callback
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.PermissionAwareActivity
import com.jumio.defaultui.JumioActivity
import com.jumio.sdk.JumioSDK
import com.jumio.sdk.credentials.JumioCredentialCategory.FACE
import com.jumio.sdk.credentials.JumioCredentialCategory.ID
import com.jumio.sdk.enums.JumioDataCenter
import com.jumio.sdk.preload.JumioPreloadCallback
import com.jumio.sdk.preload.JumioPreloader
import com.jumio.sdk.result.JumioIDResult
import com.jumio.sdk.result.JumioResult

/*
* Copyright (c) 2022. Jumio Corporation All rights reserved.
*/

class JumioModule(context: ReactApplicationContext) : JumioBaseModule(context), JumioPreloadCallback {
    companion object {
        private const val REQUEST_CODE = 101
    }

    private var preloaderFinishedCallback: Callback? = null
    private var authorizationToken: String? = null
    private var dataCenter: String? = null

    private val mActivityEventListener =
        object : BaseActivityEventListener() {
            override fun onActivityResult(activity: Activity, requestCode: Int, resultCode: Int, data: Intent?) {
                if (requestCode == REQUEST_CODE) {
                    data?.let {
                        val jumioResult = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                            it.getSerializableExtra(JumioActivity.EXTRA_RESULT, JumioResult::class.java)
                        } else {
                            @Suppress("DEPRECATION")
                            it.getSerializableExtra(JumioActivity.EXTRA_RESULT) as JumioResult?
                        }

                        if (jumioResult?.isSuccess == true) {
                            pendingResult = jumioResult
                            pendingErrorCode = null
                            pendingErrorMsg = null
                            sendScanResult(jumioResult)
                        } else {
                            sendCancelResult(jumioResult)
                        }
                    }
                }
            }
        }

    init {
        reactContext.addActivityEventListener(mActivityEventListener)
    }

    @ReactMethod
    fun isRooted(promise: Promise) = promise.resolve(JumioSDK.isRooted(reactContext))

    @ReactMethod
    fun initialize(authorizationToken: String, dataCenter: String, promise: Promise) {
        val jumioDataCenter = getJumioDataCenter(dataCenter)

        when {
            jumioDataCenter == null -> promise.reject("ERROR", "Invalid Datacenter value.")
            authorizationToken.isEmpty() -> promise.reject("ERROR", "Missing required parameters one-time session authorization token.")
            else -> {
                this.authorizationToken = authorizationToken
                this.dataCenter = dataCenter
                promise.resolve("Initialized")
            }
        }
    }

    @ReactMethod
    fun setupCustomizations(customizations: ReadableMap?) {
    }

    @ReactMethod
    fun start() {
        if (this.authorizationToken.isNullOrEmpty() || this.dataCenter.isNullOrEmpty()) {
            showErrorMessage("SDK not initialized. Please call initialize() before start().")
            return
        }

        if (!JumioSDK.hasAllRequiredPermissions(reactContext)) {
            val mp = JumioSDK.getMissingPermissions(reactContext)

            val activity = reactContext.currentActivity as? PermissionAwareActivity
            activity?.requestPermissions(mp, PERMISSION_REQUEST_CODE) { requestCode, _, grantResults ->
                if (requestCode == PERMISSION_REQUEST_CODE) {
                    if (grantResults.all { it == PackageManager.PERMISSION_GRANTED }) {
                        initSdk()
                    } else {
                        showErrorMessage("You need to grant all required permissions to continue")
                    }
                    true
                } else false
            }
        } else {
            initSdk()
        }
    }

    @ReactMethod
    fun checkCachedResult() {
        pendingResult?.let {
            sendScanResult(it)
            pendingResult = null
            return
        }
        pendingErrorMsg?.let {
            sendErrorObject(pendingErrorCode, it)
            pendingErrorCode = null
            pendingErrorMsg = null
        }
    }

    @ReactMethod
    fun setPreloaderFinishedBlock(completion: Callback) {
        with(JumioPreloader) {
            init(reactContext)
            setCallback(this@JumioModule)
        }
        preloaderFinishedCallback = completion
    }

    @ReactMethod
    fun preloadIfNeeded() {
        with(JumioPreloader) {
            preloadIfNeeded()
        }
    }

    override fun preloadFinished() {
        preloaderFinishedCallback?.invoke()
    }

    private fun initSdk() {
        try {
            val intent = Intent(reactApplicationContext.currentActivity, JumioActivity::class.java).apply {
                putExtra(JumioActivity.EXTRA_TOKEN, authorizationToken)
                putExtra(JumioActivity.EXTRA_DATACENTER, dataCenter)

                //The following intent extra can be used to customize the Theme of Default UI
                putExtra(JumioActivity.EXTRA_CUSTOM_THEME, R.style.AppThemeCustomJumio)
            }
            reactApplicationContext.currentActivity?.startActivityForResult(intent, REQUEST_CODE)
        } catch (e: Exception) {
            showErrorMessage("Error starting the Jumio SDK: " + e.localizedMessage)
        }
    }

    private fun sendScanResult(jumioResult: JumioResult?) {
        val result = buildResultMap(jumioResult)
        sendEvent(RESULT_KEY, result)
    }

    private fun buildResultMap(jumioResult: JumioResult?): WritableMap {
        val accountId = jumioResult?.accountId
        val workflowId = jumioResult?.workflowExecutionId
        val credentialInfoList = jumioResult?.credentialInfos

        val result = Arguments.createMap()
        val credentialArray = Arguments.createArray()

        credentialInfoList?.let {
            accountId?.let { result.putString("accountId", it) }
            workflowId?.let { result.putString("workflowId", it) }

            credentialInfoList.forEach {
                val eventResultMap = Arguments.createMap().apply {
                    putString("credentialId", it.id)
                    putString("credentialCategory", it.category.toString())
                }
                when (it.category) {
                    ID -> {
                        val idResult = jumioResult.getIDResult(it)
                        idResult?.let { handleIdResult(idResult, eventResultMap) }
                    }
                    FACE -> {
                        val faceResult = jumioResult.getFaceResult(it)
                        faceResult?.passed?.let { passed -> eventResultMap.putString("passed", passed.toString()) }
                    }
                    else -> {}
                }
                credentialArray.pushMap(eventResultMap)
            }
            result.putArray("credentials", credentialArray)
        }
        return result
    }

    private fun handleIdResult(idResult: JumioIDResult, eventResultMap: WritableMap) =
        with(idResult) {
            eventResultMap.apply {
                country?.let { putString("selectedCountry", it) }
                idType?.let { putString("selectedDocumentType", it) }
                idSubType?.let { putString("selectedDocumentSubType", it) }
                documentNumber?.let { putString("idNumber", it) }
                personalNumber?.let { putString("personalNumber", it) }
                issuingDate?.let { putString("issuingDate", it) }
                expiryDate?.let { putString("expiryDate", it) }
                issuingCountry?.let { putString("issuingCountry", it) }
                lastName?.let { putString("lastName", it) }
                firstName?.let { putString("firstName", it) }
                gender?.let { putString("gender", it) }
                nationality?.let { putString("nationality", it) }
                dateOfBirth?.let { putString("dateOfBirth", it) }
                address?.let { putString("addressLine", it) }
                city?.let { putString("city", it) }
                subdivision?.let { putString("subdivision", it) }
                postalCode?.let { putString("postCode", it) }
                placeOfBirth?.let { putString("placeOfBirth", it) }
                mrzLine1?.let { putString("mrzLine1", it) }
                mrzLine2?.let { putString("mrzLine2", it) }
                mrzLine3?.let { putString("mrzLine3", it) }
            }
        }

    private fun sendCancelResult(jumioResult: JumioResult?) {
        val errorCode = jumioResult?.error?.code ?: ""
        val errorMsg = jumioResult?.error?.message ?: "There was a problem extracting the scan result"

        pendingResult = null
        pendingErrorCode = errorCode
        pendingErrorMsg = errorMsg

        sendErrorObject(errorCode, errorMsg)
    }

    private fun getJumioDataCenter(dataCenter: String) = try {
        JumioDataCenter.valueOf(dataCenter)
    } catch (e: IllegalArgumentException) {
        null
    }
}